### PR TITLE
[run_glue] add speed metrics

### PR DESCRIPTION
**Pull request by stas00**
This PR starts to sync with recent changes in trainer+finetune_trainer.py

* train: (sync with finetune_trainer):
   - prints and saves train speed metrics - needed for benchmarking
   - saves the state,
* eval: sorts metrics logging info

@sgugger

### CODE REVIEW

File: examples/text-classification/run_glue.py [hunk 1]
```diff
@@ -350,11 +350,29 @@ def compute_metrics(p: EvalPrediction):
...
+                for key, value in sorted(metrics.items()):
+                    logger.info(f"  {key} = {value}")
+                    writer.write(f"{key} = {value}\n")
+
+            # Need to save the state, since Trainer.save_model saves only the tokenizer with the model
+            trainer.state.save_to_json(os.path.join(training_args.output_dir, "trainer_state.json"))
+
+            # For convenience, we also re-save the tokenizer to the same directory,
+            # so that you can share your model easily on huggingface.co/models =)
+            tokenizer.save_pretrained(training_args.output_dir)
```
Comment (sgugger) [inline examples/text-classification/run_glue.py:374]: No this is done by `trainer.save_model()`  at line 356 so there is no need to redo it.

File: examples/text-classification/run_glue.py [hunk 2]
```diff
@@ -350,11 +350,29 @@ def compute_metrics(p: EvalPrediction):
 
     # Training
     if training_args.do_train:
-        trainer.train(
+        train_result = trainer.train(
             model_path=model_args.model_name_or_path if os.path.isdir(model_args.model_name_or_path) else None
         )
+        metrics = train_result.metrics
+        metrics["train_n_objs"] = len(train_dataset)
```
Comment (sgugger) [inline examples/text-classification/run_glue.py:357]: This is logged already at the beginning of training under `num_examples`, so not sure it adds anything.

File: examples/text-classification/run_glue.py [hunk 3]
```diff
@@ -369,12 +387,13 @@ def compute_metrics(p: EvalPrediction):
 
         for eval_dataset, task in zip(eval_datasets, tasks):
             eval_result = trainer.evaluate(eval_dataset=eval_dataset)
+            eval_result["eval_n_objs"] = len(eval_dataset)
```
Comment (sgugger) [inline examples/text-classification/run_glue.py:390]: Same, not sure this adds anything (I had put them in finetune_trainer for backward compatibility reasons mostly).

**Review by sgugger — APPROVED** [review]
LGTM, thanks!

File: examples/text-classification/run_glue.py (hunk 2 above)
Comment (stas00) [reply examples/text-classification/run_glue.py:357]: it doesn't log speed, or are you referring to just `train_n_objs`?

Ideally we should refactor all these scripts to use the same format and sequence, and helper functions

File: examples/text-classification/run_glue.py (hunk 1 above)
Comment (stas00) [reply examples/text-classification/run_glue.py:374]: we are doing it in finetune_trainer - I'm just trying to sync. Let's pick one model and have others follow it.

I will remove it for now.

**Pull request merged**